### PR TITLE
fix(TDP-5590): Add preparation cleanup on dataset update

### PR DESCRIPTION
**Pull request by fhuaulme**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-5590

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [ ] The new code does not introduce new technical issues (sonar / eslint)
- [ ] Functional tests have been performed
- [ ] Docker configuration files for config-std or config-cloud profiles are impacted

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [x] No, and no need to (backend changes only)

### CODE REVIEW

File: dataprep-preparation/src/main/java/org/talend/dataprep/preparation/event/PreparationUpdateListener.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    @Autowired
+    private PreparationRepository preparationRepository;
+
+    @EventListener
+    public void onUpdate(DatasetUpdatedEvent event) {
+        final Expression filter = TqlBuilder.eq("dataSetId", event.getSource().getId());
+        preparationRepository
+                .list(Preparation.class, filter) //
+                .forEach(preparation -> {
+                    // Reset preparation row metadata.
```
[the remark this answers is not in the document]
Comment (fhuaulme) [reply dataprep-preparation/src/main/java/org/talend/dataprep/preparation/event/PreparationUpdateListener.java]: 👍

File: dataprep-preparation/src/main/java/org/talend/dataprep/preparation/event/PreparationUpdateListener.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                    preparation.setRowMetadata(event.getSource().getRowMetadata());
+                    preparationRepository.add(preparation);
+
+                    // Reset step row metadata in preparation's steps.
+                    preparationUtils
+                            .listSteps(preparation.getHeadId(), preparationRepository) //
+                            .stream() //
+                            .filter(s -> !Step.ROOT_STEP.id().equals(s.id())) //
```
[the remark this answers is not in the document]
Comment (fhuaulme) [reply dataprep-preparation/src/main/java/org/talend/dataprep/preparation/event/PreparationUpdateListener.java]: 👍

File: dataprep-preparation/src/test/java/org/talend/dataprep/preparation/event/PreparationUpdateListenerTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        when(preparationRepository.get(eq(step2.id()), eq(Step.class))).thenReturn(step2);
+
+        // when
+        final DataSetMetadata metadata = new DataSetMetadata();
+        metadata.setId("ds-1234");
+        preparationUpdateListener.onUpdate(new DatasetUpdatedEvent(metadata));
+
+        // then
+        verify(preparationRepository, times(1)).add(any(Preparation.class));
+        verify(preparationRepository, times(1)).remove(eq(StepRowMetadata.class), eq(TqlBuilder.eq("id", "srmd-1")));
```
[the remark this answers is not in the document]
Comment (fhuaulme) [reply dataprep-preparation/src/test/java/org/talend/dataprep/preparation/event/PreparationUpdateListenerTest.java]: 👍

File: dataprep-preparation/src/main/java/org/talend/dataprep/preparation/event/PreparationUpdateListener.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        final Expression filter = TqlBuilder.eq("dataSetId", event.getSource().getId());
+        preparationRepository
+                .list(Preparation.class, filter) //
+                .forEach(preparation -> {
+                    // Reset preparation row metadata.
+                    preparation.setRowMetadata(event.getSource().getRowMetadata());
+                    preparationRepository.add(preparation);
+
+                    // Reset step row metadata in preparation's steps.
+                    preparationUtils
```
Comment (aleboulanger) [inline dataprep-preparation/src/main/java/org/talend/dataprep/preparation/event/PreparationUpdateListener.java]: we could avoid several `preparationRepository#remove` and reduce to only one with: 

```java
String[] ids = preparationUtils.listSteps(preparation.getHeadId(), preparationRepository) //
                            .stream() //
                            .filter(s -> !Step.ROOT_STEP.id().equals(s.id())) //
                            .filter(s -> s.getRowMetadata() != null) //
                            .map(Step::getRowMetadata)
                            .toArray(String[]::new);
                    
                    preparationRepository.remove(StepRowMetadata.class, in("id", ids));
```

> `TqlBuilder#in` doesn't accept `Stream` but only Arrays (whereas inner method will do an Arrays.stream()...)
> don't know if make sense to create a PR for this topic

**Pull request merged**